File: top150/_02_TwoPointers/q15.java
package top150._02_TwoPointers;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class q15 {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        Set<List<Integer>> set = new HashSet<>();
        Arrays.sort(nums);
        for(int i = 0; i < nums.length-2; i++) {
            int head = i+1;
            int tail = nums.length-1;
            while(head < tail) {
                int addup = nums[i] + nums[head] + nums[tail];
                if(addup == 0) {
                    set.add(Arrays.asList(nums[i], nums[head], nums[tail]));
                    while(head < tail && nums[head] == nums[head+1]) head++;
                    while(head < tail && nums[tail] == nums[tail-1]) tail--;
                    head++;
                    tail--;
                }
                if(addup < 0) head++;
                if(addup > 0) tail--;
            }

        }
        
        res.addAll(set);
        return res;
    }
}

File: top150/_02_TwoPointers/q15.py
class Solution:
    def threeSum(self, nums):
        nums.sort()  # [-4, -1, -1, 0, 1, 2]
        # print(nums)
        n = len(nums)
        result = []
        for i in range(n):
            if nums[i] > 0:
                break
            if i == 0 or nums[i] != nums[i-1]:
                self.twoSum(nums, i, result)
        return result

    def twoSum(self, nums, i, result):
        n = len(nums)
        j = i+1
        seen = set()
        while j < n:
            # print(i,j)
            # print(nums[i], nums[j])
            remain = -nums[i] - nums[j]
            # print(remain)
            # print(seen)
            if remain in seen:
                result.append([nums[i], remain, nums[j]])
                while j+1 < n and nums[j] == nums[j+1]:
                    j += 1
            seen.add(nums[j])
            j += 1
